---
layout: page
title: About
permalink: /about/
---

We are focused on the practical knowledge and skill building to perform phylogenetic analyses that can include few or many (phylogenomic) sequences.
Mornings will cover basic statistical theory while the afternoons will focus on practical applications with population software used for real-time analyses.
Breaks are built into the day and participants are encouraged to discuss their research and learn from each other over the course of the workshop.<br>

### Learning Goals
* Understand how molecular sequence data can be used to estimate phylogenetic trees
* Understand basic features of models of molecular evolution 
* Develop strategies for estimating species trees with many loci
* Understand causes of gene tree variation
* Evaluate confidence in phylogenetic or species tree estimates
* Develop strategies for communicating and visualizing phylogenomic results

### Learning Outcomes
* Estimate maximum likelihood trees with [IQTREE](http://www.iqtree.org/)
* Estimate species trees with [BPP](https://github.com/bpp/bpp) and [ASTRAL](https://github.com/smirarab/ASTRAL)
* Evaluate node support with [quartet sampling](https://github.com/FePhyFoFum/quartetsampling)
* Produce publication quality graphics with [ggtree](https://github.com/YuLab-SMU/ggtree) in R

### Reading List
Reading of suggested papers can be valuable but is not necessary. Citations will be provided from lectures and labs. These will be updated on the [reading page]({{site.baseurl}}/reading) in the coming weeks. PDF files will be provided to participants.